indexing-hadoop/src/main/java/io/druid/indexer/DetermineHashedPartitionsJob.java

```diff
@@ -324,12 +324,9 @@ protected void reduce(
             HyperLogLogCollector.makeCollector(ByteBuffer.wrap(value.getBytes(), 0, value.getLength()))
         );
       }
-      Optional<Interval> intervalOptional = config.getGranularitySpec().bucketInterval(new DateTime(key.get()));
 
-      if (!intervalOptional.isPresent()) {
-        throw new ISE("WTF?! No bucket found for timestamp: %s", key.get());
-      }
-      Interval interval = intervalOptional.get();
+      Interval interval = config.getGranularitySpec().getSegmentGranularity().bucket(new DateTime(key.get()));
```

Comment:
Should this handle two cases of whether ```determineIntervals``` is true or not like https://github.com/druid-io/druid/blob/master/indexing-hadoop/src/main/java/io/druid/indexer/DetermineHashedPartitionsJob.java#L257-L267?

Reply:
I think @jihoonson is right -- the original code looks buggy. The new code is buggy too but in a different way. I think we fix both bugs by having an `if (determineIntervals)` branch in DetermineCardinalityReducer, where the true branch does `spec.getSegmentGranularity().bucket()` and the false branch does `spec.bucketInterval()`.

It makes sense because `spec.getSegmentGranularity().bucket()` operates without looking at the provided intervals at all, so you only want to call that when the intervals are _not_ provided. It will generate segment intervals that match the segmentGranularity exactly. When the intervals _are_ provided then `spec.bucketInterval()` respects them, so it's preferred in that case.

Reply:
@gianm i have made this change locally. i am having difficulty compiling druid locally. i'll push it up once that is resolved.